af/fastpath/fastpath/core.py

```diff
@@ -0,0 +1,574 @@
+    """Fetch measurements from S3 and the collectors
+    """
+    # no --start-day or --end-day   -> Run over SSH
+    # --start-day                   -> Run on old cans, then over SSH
+    # --start-day and --end-day     -> Run on old cans
+    # --start-day and --end-day  with the same date     -> NOP
+    today = date.today()
+    if start_day and start_day < today:
+        # TODO: fetch day N+1 while processing day N
+        log.info("Fetching older cans from S3")
+        day = start_day
+        while day < (end_day or today):
+            log.info("Processing %s", day)
+            s3feeder.fetch_cans_for_a_day_with_cache(conf, day)
+            for measurement_tup in load_s3_reports(day):
+                yield measurement_tup
+
+            day += timedelta(days=1)
+
+        if end_day:
+            return
+
+    ## Fetch measurements from collectors: backlog and then realtime ##
+    log.info("Starting fetching over SSH")
+    for measurement_tup in sshfeeder.feed_measurements_from_collectors(conf):
+        yield measurement_tup
+
+
+@metrics.timer("match_fingerprints")
+def match_fingerprints(measurement):
+    """Match fingerprints against HTTP headers and bodies.
+    """
+    # TODO: apply only on web_connectivity
+    msm_cc = measurement["probe_cc"]
+    if msm_cc not in fingerprints:
+        return []
+
+    matches = []
+    for req in measurement["test_keys"].get("requests", ()):
+        r = req.get("response", None)
+        if r is None:
+            continue
+
+        # Match HTTP body if found
+        body = r["body"]
+        if body is not None:
+            for fp in fingerprints[msm_cc].get("body_match", []):
+                # fp: {"body_match": "...", "locality": "..."}
+                tb = time.time()
+                bm = fp["body_match"]
+                if bm in body:
+                    matches.append(fp)
+                    log.debug("matched body fp %s %r", msm_cc, bm)
+
+                per_s("fingerprints_bytes", len(body), tb)
+
+        del (body)
+
+        # Match HTTP headers if found
+        headers = r.get("headers", {})
+        if not headers:
+            continue
+        headers = {h.lower(): v for h, v in headers.items()}
+        for fp in fingerprints[msm_cc].get("header_full", []):
+            name = fp["header_name"]
+            if name in headers and headers[name] == fp["header_full"]:
+                matches.append(fp)
+                log.debug("matched header full fp %s %r", msm_cc, fp["header_full"])
+
+        for fp in fingerprints[msm_cc].get("header_prefix", []):
+            name = fp["header_name"]
+            prefix = fp["header_prefix"]
+            if name in headers and headers[name].startswith(prefix):
+                matches.append(fp)
+                log.debug("matched header prefix %s %r", msm_cc, prefix)
+
+    return matches
+
+
+@metrics.timer("score_measurement")
+def score_measurement(msm, matches):
```

Comment:
Maybe this should be called `score_web_connectivity`

Reply:
The idea is to extend it to handle any msmt.